doc/developer-guides/modularity.rst

```diff
@@ -0,0 +1,89 @@
+.. _modularity:
+
+ACRN Hypervisor: Modular Design
+###############################
+
+Overview
+********
+
+ACRN highly emphasizes modular design, i.e. the separation of functionality info
+modules that define a concise set of interfaces. The goals of modular design
+include:
+
+* **Understandability** A modular design is easier to understand due to
+  encapsulation.
+* **Testability** Modules can be integrated and tested in the reverse order of
+  dependencies among them. White-box integration tests help improve the coverage
+  of tests and identify corner cases that are hard to trigger when testing the
+  hypervisor as a whole.
+* **Configurability** Modular design makes it easy to configure certain
+  functionalities in or out. This is crucial in safety-critical scenarios
+  because absence of irrelevant code is required in both MISRA-C and functional
+  safety standards.
+* **Meet functional safety requirements** Functional safety standards explicitly
+  requires a hierarchical structure of modules in software architectural
+  design. This applies to any safety integrity level defined in IEC 61508
+  [IEC 61508-3]_ and ISO 26262 [ISO 26262-6]_.
+
+Principles
+**********
+
+* Each source file shall belong to one module only. One module may consist of
+  one or multiple source files, though. A source file can be a C source, a C
+  header, an assembly file, or any other file that is built into the binary
+  image.
+* Each module shall have well-defined interfaces, including the exported
+  functions and global variables. Functions and variables that are not
+  interfaces shall be static and used inside the module only.
+* Dependencies among the modules should be acyclic. Any cyclic dependency must
+  be deviated explicitly.
+* The complexity of a module shall be limited.
+
+Architecture
+************
+
+The following figure shows the high-level components of ACRN hypervisor. ACRN
+hypervisor adopts a layered design where higher layers can invoke the interfaces
+of lower layers but not vice versa. The only exception is the invocation of
+initialization routine in the **Boot** component.
```

Comment:
Can you explain a bit more about this exception?

Reply:
Added more description on this exception, why it is introduced and how to work it around during integration testing.